## **Data Preparation**

In [58]:
import pandas as pd 

df=pd.read_excel('https://data.go.th/dataset/5370f16a-b553-45ca-929d-7db9a6fc5043/resource/48039a2a-2f01-448c-b2a2-bb0d541dedcd/download/tambon.xlsx')
tambon_df=df.groupby('TA_ID').mean().join(df[['TA_ID','TAMBON_E']].set_index('TA_ID'))[['LAT','LONG','TAMBON_E']]
tambon_df.to_csv('tambon_location.csv')

/var/folders/zn/20g6j7t91810cl_ts2h592y40000gn/T/ipykernel_53361/3303041387.py:4: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  tambon_df=df.groupby('TA_ID').mean().join(df[['TA_ID','TAMBON_E']].set_index('TA_ID'))[['LAT','LONG','TAMBON_E']]


In [107]:
tambon_df[:50]

LAT     LONG                    TAMBON_E
TA_ID                                              
100101  13.751  100.492  Phra Borom Maha Ratchawang
100102  13.744  100.499         Wang Burapha Phirom
100103  13.750  100.499            Wat Ratchabophit
100104  13.751  100.503                  Samran Rat
100105  13.754  100.497            San Chaopho Suea
100106  13.753  100.500                Sao Chingcha
100107  13.757  100.501                 Bowon Niwet
100108  13.760  100.498                   Talat Yot
100109  13.762  100.495             Chana Songkhram
100110  13.762  100.503               Ban Phan Thom
100111  13.765  100.505             Bang Khun Phrom
100112  13.768  100.499              Wat Sam Phraya
100201  13.772  100.513                       Dusit
100202  13.778  100.505            Wachira Phayaban
100203  13.767  100.520              Suan Chit Lada
100204  13.758  100.517            Si Yaek Maha Nak
100206  13.789  100.522                         NaN
100301  13.823  100.897                         NaN
100302  13.869  100.885                         NaN
100303  13.914  100.820                         NaN
100304  13.914  100.880                         NaN
100305  13.840  100.836                         NaN
100306  13.872  100.810                         NaN
100307  13.797  100.844                         NaN
100308  13.781  100.877                         NaN
100401  13.734  100.520                         NaN
100402  13.730  100.525                         NaN
100403  13.724  100.530                         NaN
100404  13.727  100.527                         NaN
100405  13.725  100.514                         NaN
100502  13.868  100.606                         NaN
100508  13.866  100.650                         NaN
100601  13.786  100.635                         NaN
100608  13.756  100.660                         NaN
100701  13.744  100.520                         NaN
100702  13.742  100.526                         NaN
100703  13.740  100.535                         NaN
100704  13.736  100.546                         NaN
100801  13.743  100.514                         NaN
100802  13.749  100.512                         NaN
100803  13.753  100.513                         NaN
100804  13.752  100.507                         NaN
100805  13.759  100.511                         NaN
100905  13.692  100.626                         NaN
101001  13.804  100.723                         NaN
101002  13.812  100.774                         NaN
101101  13.723  100.771                         NaN
101102  13.753  100.723                         NaN
101103  13.753  100.754                         NaN
101104  13.765  100.817                         NaN

### Choose capacity and number of tambons

In [108]:
capacity = 150
n = 50

In [109]:
import random 

random.seed(31)

tambons=list(tambon_df.index)
random.shuffle(tambons)
tambons=tambons[:n]
tambons=[[tambon,(int(random.random()*6+5)*10)] for tambon in tambons]
tambons[0][1]=0

In [110]:
tambons

[[400109, 0],
 [730114, 90],
 [101602, 100],
 [320701, 100],
 [190112, 60],
 [330203, 50],
 [571701, 100],
 [210202, 80],
 [600906, 60],
 [551405, 80],
 [960804, 100],
 [471208, 100],
 [650107, 70],
 [470305, 100],
 [130703, 90],
 [550904, 50],
 [720204, 60],
 [390309, 60],
 [140902, 50],
 [850303, 90],
 [650905, 100],
 [141109, 100],
 [640510, 50],
 [910106, 50],
 [710403, 60],
 [960203, 50],
 [860706, 70],
 [801401, 100],
 [120604, 90],
 [140109, 80],
 [100110, 90],
 [710509, 90],
 [840401, 70],
 [500903, 50],
 [340108, 60],
 [440707, 100],
 [311305, 60],
 [760123, 100],
 [510101, 100],
 [660408, 90],
 [810302, 100],
 [600601, 50],
 [190307, 70],
 [401707, 70],
 [570701, 60],
 [450211, 80],
 [650210, 100],
 [191301, 80],
 [330122, 70],
 [640607, 70]]

## **Your Tasks**

Write a program which can create a route that a truck can deliver items to all places stored in `tambons` list. The number of items that the truck must deliver is the second element in the list. 

The truck has capacity as specified in the variable `capacity`. When it deliver all items, it must go back to the first location in the list to load the items and start deliver again. The process will repeatly be done until all items are delivered. 

For example, if the data in tambons are: 

```
[[400109, 0],
 [730114, 90],
 [101602, 100],
 [320701, 100],
 [190112, 60],
 [330203, 50],
 [571701, 100]]
 ```

 The starting point is tambon `400109` and if the capacity is `300`. One route can be `[400109,101602,320701,571701,400109,730114,190112,330203,400109]` which means 

 1. load `300 (100+100+100)` items and deliver to `101602,320701,571701` respectively
 2. go back to starting point `400109`
 3. load `200 (90+60+50)` items and deliver to `730114,190112,330203`
 4. go back to starting point 



 ### **Scoring scriteria** (from 50 marks)

 1. Use only techniques and codes from the class (0 marks for codes from other resources)
 2. Can find the route to all locations and total distance (in km.) (35 marks)
 3. The route is shortest (40 marks) 
 4. Can find the route which follows the capacity criteria (not neccessary to be the shortest (45 marks)
 5. Can find the shortest route which follow the capacity criteria (50 marks)

### Find km from lat and long 

```
from geopy.distance import geodesic

lat1, lon1 = 40.6892, -74.0445
lat2, lon2 = 37.7749, -122.4194
distance = geodesic((lat1, lon1), (lat2, lon2)).km
```

### Get lat and long from tambon_df 

```
tambon_df.loc[tambon_id]
```

For example,
```
print(tambon_df.loc[571701])
```

# Main

In [111]:
from geopy.distance import geodesic

In [118]:
class DeliveryRoute:
    def __init__(self,n, capacity, tambons):
        self.n = n
        self.capacity = capacity
        self.tambons = tambons
        self.total_distance = 0
        self.current_capacity = capacity
        self.start_tambon = tambons[0][0]
        self.visited_tambons = [tambons[0][0]]
        self.unvisited_tambons = [tambon for tambon in tambons[1:]]

    def get_distance(self, tambon1, tambon2):
        print(f'tambon1 : {tambon1}, tambon2 : {tambon2}')
        lat1, long1 = tambon_df.loc[tambon1][['LAT','LONG']]
        lat2, long2 = tambon_df.loc[tambon2][['LAT','LONG']]
        print(f'Old type: {tambon1} {lat1} {long1}, {tambon2} {lat2} {long2}')
        # convert to float
        lat1 = float(lat1)
        long1 = float(long1)
        lat2 = float(lat2)
        long2 = float(long2)
        print(f'Lat1 : {lat1} {type(lat1)}, Long1 : {long1} {type(long1)}, Lat2 : {lat2} {type(lat2)}, Long2 : {long2} {type(long2)}')
        return geodesic((float(lat1), float(long1)), (float(lat2), float(long2))).km
    
    def get_next_tambon_idx(self):
        for idx in range(len(self.unvisited_tambons)):
            load = self.unvisited_tambons[idx][1]
            if self.current_capacity >= load:
                return idx
        return -1
    
    def find_route(self):
        while True:
            if len(self.visited_tambons) == 0:
                print('No solution')
                break
            if len(self.unvisited_tambons) == 0:
                print('Route found!!!')
                distance = self.get_distance(self.visited_tambons[-1], self.start_tambon)
                self.total_distance += distance
                self.visited_tambons.append(self.start_tambon)
                print('Total distance: ', self.total_distance)
                print('Visited tambons: ', self.visited_tambons)
                break
            # Sort unvisited tambons by its load from high to low
            self.unvisited_tambons = sorted(self.unvisited_tambons, key=lambda x: x[1], reverse=True)
            next_tambon_idx = self.get_next_tambon_idx()
            if next_tambon_idx == -1:
                distance = self.get_distance(self.visited_tambons[-1], self.start_tambon)
                self.total_distance += distance
                self.current_capacity = self.capacity
                self.visited_tambons.append(self.start_tambon)
            else:
                self.current_capacity -= self.unvisited_tambons[next_tambon_idx][1]
                distance = self.get_distance(self.visited_tambons[-1], self.unvisited_tambons[next_tambon_idx][0])
                self.total_distance += distance
                self.visited_tambons.append(self.unvisited_tambons.pop(next_tambon_idx)[0])
                
                

In [119]:
# This time find route with the shortest distance
class ShortestDeliveryRoute(DeliveryRoute):
    def get_next_tambon_idx(self):
        min_distance = 100000000
        min_idx = -1
        for idx in range(len(self.unvisited_tambons)):
            load = self.unvisited_tambons[idx][1]
            if self.current_capacity >= load:
                distance = self.get_distance(self.visited_tambons[-1], self.unvisited_tambons[idx][0])
                if distance < min_distance:
                    min_distance = distance
                    min_idx = idx
        return min_idx
    def find_route(self):
        while True:
            if len(self.visited_tambons) == 0:
                print('No solution')
                break
            if len(self.unvisited_tambons) == 0:
                print('Route found!!!')
                distance = self.get_distance(self.visited_tambons[-1], self.start_tambon)
                self.total_distance += distance
                self.visited_tambons.append(self.start_tambon)
                print('Total distance: ', self.total_distance)
                print('Visited tambons: ', self.visited_tambons)
                break
            # Sort unvisited tambons by its load from high to low
            self.unvisited_tambons = sorted(self.unvisited_tambons, key=lambda x: x[1], reverse=True)
            next_tambon_idx = self.get_next_tambon_idx()
            if next_tambon_idx == -1:
                distance = self.get_distance(self.visited_tambons[-1], self.start_tambon)
                self.total_distance += distance
                self.current_capacity = self.capacity
                self.visited_tambons.append(self.start_tambon)
            else:
                self.current_capacity -= self.unvisited_tambons[next_tambon_idx][1]
                distance = self.get_distance(self.visited_tambons[-1], self.unvisited_tambons[next_tambon_idx][0])
                self.total_distance += distance
                self.visited_tambons.append(self.unvisited_tambons.pop(next_tambon_idx)[0])
            
    

In [120]:
deliverRoute = DeliveryRoute(n, capacity, tambons)
deliverRoute.find_route()

tambon1 : 400109, tambon2 : 101602
Lat1 : 16.384 <class 'float'>, Long1 : 102.713 <class 'float'>, Lat2 : 13.733 <class 'float'>, Long2 : 100.475 <class 'float'>
tambon1 : 101602, tambon2 : 330203
Lat1 : 13.733 <class 'float'>, Long1 : 100.475 <class 'float'>, Lat2 : 15.223 <class 'float'>, Long2 : 104.36 <class 'float'>
tambon1 : 330203, tambon2 : 400109
Lat1 : 15.223 <class 'float'>, Long1 : 104.36 <class 'float'>, Lat2 : 16.384 <class 'float'>, Long2 : 102.713 <class 'float'>
tambon1 : 400109, tambon2 : 320701
Lat1 : 16.384 <class 'float'>, Long1 : 102.713 <class 'float'>, Lat2 : 15.311 <class 'float'>, Long2 : 103.843 <class 'float'>
tambon1 : 320701, tambon2 : 550904
Lat1 : 15.311 <class 'float'>, Long1 : 103.843 <class 'float'>, Lat2 : 19.338 <class 'float'>, Long2 : 100.925 <class 'float'>
tambon1 : 550904, tambon2 : 400109
Lat1 : 19.338 <class 'float'>, Long1 : 100.925 <class 'float'>, Lat2 : 16.384 <class 'float'>, Long2 : 102.713 <class 'float'>
tambon1 : 400109, tambon2 : 57

ValueError: could not convert string to float: 'LAT'

[400109,101602,320701,571701,400109,730114,190112,330203,400109] 

In [ ]:
ShortestDeliveryRoute(n, capacity, tambons).find_route()

Route found!!!
Total distance:  5178.087862483448
Visited tambons:  [400109, 320701, 330203, 400109, 600906, 190112, 400109, 551405, 400109, 101602, 400109, 730114, 400109, 210202, 400109, 571701, 400109]
